File: verify/claims.js
// Independent recomputation of the numbers on the front page, in Node.
//
// Two things happen here. Every phrase in verify/claims.tsv must appear in the
// rendered text of index.html exactly once, and every published value must be
// reproducible from the source values by the stated rule. The page is the only
// place these numbers are written out in prose, so nothing else would notice if
// a hand edit changed one.
//
// Also checks the JSON-LD block with a real JSON parser and requires the
// identity URL in it to match the canonical link and the sitemap entry.

const fs = require("fs");
const path = require("path");

const root = path.resolve(__dirname, "..");
const html = fs.readFileSync(path.join(root, "index.html"), "utf8");
const tsv = fs.readFileSync(path.join(root, "verify", "claims.tsv"), "utf8");

let bad = 0;
const fail = (m) => { console.log("FAIL " + m); bad++; };

// Visible text: drop script and style blocks, drop tags, unescape the few
// entities this page uses, collapse whitespace.
const text = html
  .replace(/<script[\s\S]*?<\/script>/gi, " ")
  .replace(/<style[\s\S]*?<\/style>/gi, " ")
  .replace(/<[^>]+>/g, " ")
  .replace(/&amp;/g, "&").replace(/&lt;/g, "<").replace(/&gt;/g, ">")
  .replace(/\s+/g, " ")
  .trim();

const rows = tsv.trim().split("\n").map((l) => l.split("\t"));
const head = rows.shift();
const col = {};
head.forEach((h, i) => { col[h] = i; });
for (const need of ["id", "rule", "a", "b", "published", "phrase"]) {
  if (!(need in col)) { fail("claims.tsv has no column " + need); }
}
if (bad) { process.exit(1); }

const round = (x, n) => {
  const f = Math.pow(10, n);
  return Math.sign(x) * Math.round(Math.abs(x) * f) / f;
};

const recompute = (rule, a, b) => {
  const m = /^(copy|round|ratio|diff)(\d*)$/.exec(rule);
  if (!m) { return null; }
  const n = m[2] === "" ? 0 : parseInt(m[2], 10);
  if (m[1] === "copy") { return a; }
  if (m[1] === "round") { return round(a, n); }
  if (m[1] === "ratio") { return round(a / b, n); }
  return round(a - b, n);
};

let phrases = 0, values = 0;
let worst = 0, worstId = "";
for (const r of rows) {
  const id = r[col.id];
  const phrase = r[col.phrase];
  const occurrences = text.split(phrase).length - 1;
  if (occurrences !== 1) {
    fail(id + ": phrase appears " + occurrences + " times in index.html, want 1: " + phrase);
  } else {
    phrases++;
  }

  const a = Number(r[col.a]);
  const b = r[col.b] === "-" ? NaN : Number(r[col.b]);
  const published = Number(r[col.published]);
  if (!Number.isFinite(a) || !Number.isFinite(published)) {
    fail(id + ": a or published is not a finite number");
    continue;
  }
  const got = recompute(r[col.rule], a, b);
  if (got === null) { fail(id + ": unknown rule " + r[col.rule]); continue; }
  const err = Math.abs(got - published);
  const tol = Math.abs(published) * 1e-12 + 1e-12;
  if (err > tol) {
    fail(id + ": rule " + r[col.rule] + " gives " + got + ", page publishes " + published);
  } else {
    values++;
    if (err > worst) { worst = err; worstId = id; }
  }
}

// JSON-LD, canonical and sitemap must agree on one URL.
const ld = /<script type="application\/ld\+json">([\s\S]*?)<\/script>/.exec(html);
if (!ld) {
  fail("no JSON-LD block in index.html");
} else {
  let obj = null;
  try { obj = JSON.parse(ld[1]); } catch (e) { fail("JSON-LD does not parse: " + e.message); }
  if (obj) {
    const canonical = /<link rel="canonical" href="([^"]+)"/.exec(html);
    const sitemap = fs.readFileSync(path.join(root, "sitemap.xml"), "utf8");
    const loc = /<loc>([^<]+)<\/loc>/.exec(sitemap);
    if (!canonical) { fail("no canonical link"); }
    else if (!loc) { fail("no <loc> in sitemap.xml"); }
    else if (obj.url !== canonical[1] || obj.url !== loc[1]) {
      fail("JSON-LD url " + obj.url + ", canonical " + canonical[1] + ", sitemap " + loc[1] + " do not agree");
    } else {
      console.log("JSON-LD parses; url, canonical and sitemap loc all " + obj.url);
    }
  }
}

if (bad) {
  console.log("JavaScript: " + bad + " failures");
  process.exit(1);
}
console.log("JavaScript: " + phrases + " phrases found once each, " + values +
  " published values reproduced, largest error " +
  (worst === 0 ? "0.0e+00" : worst.toExponential(1) + " on " + worstId));

File: verify/claims.c
/* Recompute every published value in verify/claims.tsv in C.
 *
 * Same arithmetic as the SQL, Go and JavaScript versions, written against a
 * different rounding implementation and a different TSV reader. Columns are
 * resolved by name from the header row, so reordering the file cannot silently
 * change what is compared.
 *
 * Build: cc -std=c99 -O2 -o claimc verify/claims.c -lm
 * Run:   ./claimc <repo root>
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#define MAXCOL 32
#define MAXLINE 4096

static int find_col(char *header[], int n, const char *name) {
    for (int i = 0; i < n; i++) {
        if (strcmp(header[i], name) == 0) return i;
    }
    fprintf(stderr, "claims.tsv has no column named %s\n", name);
    exit(1);
}

/* Split a tab separated line in place. Returns the field count. */
static int split(char *line, char *out[], int max) {
    int n = 0;
    char *p = line;
    out[n++] = p;
    while (*p) {
        if (*p == '\t') {
            *p = '\0';
            if (n >= max) { fprintf(stderr, "too many columns\n"); exit(1); }
            out[n++] = p + 1;
        }
        p++;
    }
    return n;
}

static double round_to(double x, int digits) {
    double f = pow(10.0, (double)digits);
    return round(x * f) / f;
}

int main(int argc, char **argv) {
    const char *root = argc > 1 ? argv[1] : ".";
    char path[1024];
    snprintf(path, sizeof path, "%s/verify/claims.tsv", root);
    FILE *f = fopen(path, "r");
    if (!f) { fprintf(stderr, "cannot open %s\n", path); return 1; }

    char line[MAXLINE];
    if (!fgets(line, sizeof line, f)) { fprintf(stderr, "empty claims.tsv\n"); return 1; }
    line[strcspn(line, "\r\n")] = '\0';
    char *header[MAXCOL];
    int ncol = split(line, header, MAXCOL);

    int c_id = find_col(header, ncol, "id");
    int c_rule = find_col(header, ncol, "rule");
    int c_a = find_col(header, ncol, "a");
    int c_b = find_col(header, ncol, "b");
    int c_pub = find_col(header, ncol, "published");

    int rows = 0, bad = 0;
    double worst = 0.0;
    char worst_id[128] = "none";

    while (fgets(line, sizeof line, f)) {
        line[strcspn(line, "\r\n")] = '\0';
        if (line[0] == '\0') continue;
        char *fld[MAXCOL];
        int n = split(line, fld, MAXCOL);
        if (n != ncol) {
            printf("FAIL ragged row %d: %d fields, header has %d\n", rows + 1, n, ncol);
            bad++; rows++; continue;
        }
        rows++;

        char *end;
        double a = strtod(fld[c_a], &end);
        if (end == fld[c_a] || *end != '\0') {
            printf("FAIL %s: column a is not a number: %s\n", fld[c_id], fld[c_a]); bad++; continue;
        }
        double pub = strtod(fld[c_pub], &end);
        if (end == fld[c_pub] || *end != '\0') {
            printf("FAIL %s: column published is not a number: %s\n", fld[c_id], fld[c_pub]); bad++; continue;
        }
        int has_b = strcmp(fld[c_b], "-") != 0;
        double b = 0.0;
        if (has_b) {
            b = strtod(fld[c_b], &end);
            if (end == fld[c_b] || *end != '\0') {
                printf("FAIL %s: column b is not a number: %s\n", fld[c_id], fld[c_b]); bad++; continue;
            }
        }
        if (!isfinite(a) || !isfinite(pub) || (has_b && !isfinite(b))) {
            printf("FAIL %s: non finite value in the row\n", fld[c_id]); bad++; continue;
        }

        const char *rule = fld[c_rule];
        double got;
        if (strcmp(rule, "copy") == 0) {
            got = a;
        } else if (strncmp(rule, "round", 5) == 0) {
            got = round_to(a, atoi(rule + 5));
        } else if (strncmp(rule, "ratio", 5) == 0) {
            if (!has_b || b == 0.0) {
                printf("FAIL %s: rule %s needs a non zero b\n", fld[c_id], rule); bad++; continue;
            }
            got = round_to(a / b, atoi(rule + 5));
        } else if (strncmp(rule, "diff", 4) == 0) {
            if (!has_b) {
                printf("FAIL %s: rule %s needs a b\n", fld[c_id], rule); bad++; continue;
            }
            got = round_to(a - b, atoi(rule + 4));
        } else {
            printf("FAIL %s: unknown rule %s\n", fld[c_id], rule); bad++; continue;
        }

        double err = fabs(got - pub);
        double tol = fabs(pub) * 1e-12 + 1e-12;
        if (err > tol) {
            printf("FAIL %s: rule %s gives %.10g, page publishes %.10g\n", fld[c_id], rule, got, pub);
            bad++;
        } else if (err > worst) {
            worst = err;
            snprintf(worst_id, sizeof worst_id, "%s", fld[c_id]);
        }
    }
    fclose(f);

    if (rows == 0) { printf("FAIL claims.tsv has no rows\n"); return 1; }
    if (bad) { printf("C: %d of %d rows disagree\n", bad, rows); return 1; }
    printf("C: %d published values reproduced, largest error %.1e on %s\n", rows, worst, worst_id);
    return 0;
}
